java/debugger/impl/src/com/intellij/debugger/memory/ui/InstancesView.java

```diff
@@ -388,4 +415,68 @@ public JavaReferenceInfo get(int index) {
       return myRefs.get(index);
     }
   }
+
+  private interface InstancesViewRepresentation {
+    void customizeView(@NotNull InstancesView view, @NotNull XDebugSession session, String className);
+
+    List<JavaReferenceInfo> fetchInstances(@NotNull EvaluationContextImpl evaluationContext);
+  }
+
+  private final class TreeRepresentation implements InstancesViewRepresentation {
+    @Override
+    public void customizeView(@NotNull InstancesView view, @NotNull XDebugSession session, String className) {
+      final StackFrameList list = new StackFrameList(myDebugProcess);
+      list.addListSelectionListener(e -> list.navigateToSelectedValue(false));
+      new DoubleClickListener() {
+        @Override
+        protected boolean onDoubleClick(@NotNull MouseEvent event) {
+          list.navigateToSelectedValue(true);
+          return true;
+        }
+      }.installOn(list);
+
+      final InstancesWithStackFrameView instancesWithStackFrame = new InstancesWithStackFrameView(
+        session, myInstancesTree, list, className
+      );
+      view.add(instancesWithStackFrame.getComponent(), BorderLayout.CENTER);
+    }
+
+    @Override
+    public List<JavaReferenceInfo> fetchInstances(@NotNull EvaluationContextImpl evaluationContext) {
+      final int limit = DebuggerUtils.isAndroidVM(myDebugProcess.getVirtualMachineProxy().getVirtualMachine())
+                        ? AndroidUtil.ANDROID_INSTANCES_LIMIT
+                        : DEFAULT_INSTANCES_LIMIT;
+      List<JavaReferenceInfo> instances = ContainerUtil.map(
+        getInstancesProvider().getInstances(limit + 1),
+        referenceInfo -> ((JavaReferenceInfo)referenceInfo)
+      );
+
+      if (instances.size() > limit) {
+        myWarningMessageConsumer.accept(XDebuggerBundle.message("memory.view.instances.warning.not.all.loaded", limit));
+        instances = instances.subList(0, limit);
+      }
+
+      return instances;
+    }
+  }
+
+
+  private final class TreeTableRepresentation implements InstancesViewRepresentation {
+    @Override
+    public void customizeView(@NotNull InstancesView view, @NotNull XDebugSession session, String className) {
+      InstancesViewTreeTableModel treeTableModel = new InstancesViewTreeTableModel(myInstancesTree);
+      JBTreeTable treeTable = new JBTreeTable(treeTableModel, myInstancesTree);
```

Comment:
Is it critical to pass `myInstancesTree ` as a tree parameter to `JBTreeTable`? As I can see, `JBTreeTable` relies on the model, so delegating via `InstancesViewTreeTableModel` looks enough